schedutils/coreset.c

```diff
@@ -0,0 +1,287 @@
+	fprintf(out, _("Core scheduling is available in kernels starting with v5.14.\n"));
+
+	fprintf(out, USAGE_MAN_TAIL("coreset(1)"));
+	exit(EXIT_SUCCESS);
+}
+
+static void print_cookie(struct coreset *cs, int isnew)
+{
+	char *msg;
+
+	msg = isnew ? _("pid %d's new cookie: 0x%0x\n") :
+		_("pid %d's current cookie: 0x%0x\n");
+
+	printf(msg, cs->pid ? cs->pid : getpid(), cs->cookie);
+}
+
+static void __attribute__((__noreturn__)) err_cookie(pid_t pid, int set)
+{
+	char *msg;
+
+	if(set == CORE_COPY)
+		msg = _("failed to copy pid %d's core scheduling cookie");
+	else
+		msg = set ? _("failed to set pid %d's core scheduling cookie") :
+			    _("failed to get pid %d's core scheduling cookie");
+
+	err(EXIT_FAILURE, msg, pid ? pid : getpid());
+}
+
+static unsigned long get_cookie(struct coreset *cs)
+{
+	unsigned long cookie;
+	pid_t pid = (cs->cmd == CORE_COPY ? 0: cs->pid); /* with copy we want to report current's cookie */
+
+	if (prctl(PR_SCHED_CORE, PR_SCHED_CORE_GET, pid, PR_SCHED_CORE_SCOPE_THREAD, &cookie) < 0 ) {
+		err_cookie(pid ? pid : getpid(), FALSE);
+	}
+	return cookie;
+}
+
+static void do_coreset(struct coreset *cs)
+{
+	/* read the current cookie */
+	cs->cookie = get_cookie(cs);
+	print_cookie(cs, FALSE);
+
+	switch (cs->cmd) {
+	case CORE_SHOW:
+		return;
+	case CORE_CREATE:
+                /* create a new cookie for given task (may be 0). Scope only applies with existing pid */
+		if (prctl(PR_SCHED_CORE, PR_SCHED_CORE_CREATE, cs->pid, cs->scope, NULL) < 0)
+			err_cookie(cs->pid, cs->cmd);
+	        break;
+	case CORE_COPY:
+		/* copy cookie, which could be none, from source pid to current task.  Scope must be 0 so we force it*/
+		if (prctl(PR_SCHED_CORE, PR_SCHED_CORE_SHARE_FROM, cs->pid, PR_SCHED_CORE_SCOPE_THREAD, NULL) < 0)
+			err_cookie(cs->pid, cs->cmd);
+		break;
+	case CORE_PUSH:
+                /* push current task's cookie, which could be none, to given pid. Scope is meaningful */
+		if (prctl(PR_SCHED_CORE, PR_SCHED_CORE_SHARE_TO, cs->pid, cs->scope, NULL) < 0)
+			err_cookie(cs->pid, cs->cmd);
+		break;
+	}
+
+	/* re-read the cookie */
+	cs->cookie = get_cookie(cs);
+	print_cookie(cs, TRUE);
+}
+
+int main(int argc, char **argv)
+{
+	pid_t pid = 0;
+	int c, copy = 0, create = 0, push = 0;
+	int do_exec = 0;
+	struct coreset cs;
+
+	static const struct option longopts[] = {
+		{ "copy",	0, NULL, 'c' },
+		{ "new",	0, NULL, 'n' },
+		{ "pid",	0, NULL, 'p' },
+		{ "scope",	0, NULL, 's' },
+		{ "to",	        0, NULL, 't' },
+		{ "help",	0, NULL, 'h' },
+		{ "version",	0, NULL, 'V' },
+		{ NULL,		0, NULL,  0  }
+	};
+
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+	close_stdout_atexit();
+
+	memset(&cs, 0, sizeof(cs));
+
+	while ((c = getopt_long(argc, argv, "+cnp:s:thV", longopts, NULL)) != -1) {
+		switch (c) {
+		case 'c':
+			copy = 1;
+			break;
+		case 'n':
+		        create = 1;
+			break;
+		case 'p':
+			pid = strtos32_or_err(optarg, _("invalid PID argument"));
```

Comment:
`strtopid_or_err`

Reply:
Cool. Did not know that existed.  There's probably a lot of stuff in the util-linux codebase that I have no idea about...